support/cas-server-support-token-tickets/src/test/java/org/apereo/cas/token/authentication/principal/TokenWebApplicationServiceResponseBuilderTests.java

```diff
@@ -139,4 +140,19 @@ public void verifyTokenBuilder() throws Exception {
             assertNotNull(JWTParser.parse(ticket));
         }
     }
+    
+    @Test
+    public void verifyTokenBuilderWhenGatewayIsTrue() throws Exception {
+        val data = "yes\ncasuser";
+        try (val webServer = new MockWebServer(8281,
+            new ByteArrayResource(data.getBytes(StandardCharsets.UTF_8), "REST Output"), MediaType.APPLICATION_JSON_VALUE)) {
+            webServer.start();
+
+            val result = responseBuilder.build(CoreAuthenticationTestUtils.getWebApplicationService("jwtservice"),
+                StringUtils.EMPTY,
+                CoreAuthenticationTestUtils.getAuthentication());
+            assertNotNull(result);
+            assertFalse(result.getAttributes().containsKey(CasProtocolConstants.PARAMETER_TICKET));
+        }
```

Comment:
How does this test verify gateway=true?

Reply:
If the service is configured to receive à service ticket as a JWT and that the request parameter gateway=true, then there is no service ticket generated. So the second parameter of the method responseBuilder.build() is empty. I renamed the test method to avoid any misunderstanding.

Reply:
There is nothing in your test about gateway=true. You need to have two tests; one with and one without that parameter.

Reply:
Just to be sure that there is no misunderstanding, let's focus on these lines:

TokenWebApplicationServiceResponseBuilder:
`if (!tokenAsResponse) {
LOGGER.debug("Registered service [{}] is not configured to issue JWTs for service tickets. "
                    + "Make sure the service property [{}] is defined and is set to true", registeredService,
                RegisteredServiceProperties.TOKEN_AS_SERVICE_TICKET.getPropertyName());
            return super.buildInternal(service, parameters);
 }`

TokenWebApplicationServiceResponseBuilderTests:
`responseBuilder.build(CoreAuthenticationTestUtils.getWebApplicationService("jwtservice"), "ST-123456",CoreAuthenticationTestUtils.getAuthentication());`

Scenario 1: The service is not configured to send the  service ticket as a JWT and there is no SSO session (no TGT, no Service ticket), gateway=false. Then:
The call to the responseBuilder looks like something like this: `responseBuilder.build(CoreAuthenticationTestUtils.getWebApplicationService("jwtservice"), "",CoreAuthenticationTestUtils.getAuthentication());` because there is no service ticket.
And as tokenAsResponse=false => execute `super.buildInternal(service, parameters)`

Scenario 2: The service is configured to send the  service ticket as a JWT and there is a SSO session with a TGT and a service ticket, gateway=false. Then:
The call to the responseBuilder looks like something like that:
`responseBuilder.build(CoreAuthenticationTestUtils.getWebApplicationService("jwtservice"),
                "ST-123456",
                CoreAuthenticationTestUtils.getAuthentication());` because there is a service ticket.
And as tokenAsResponse=true=> let's generate the JWT from the service ticket.

Scenario 3: The service is configured to send the  service ticket as a JWT and there is a SSO session with a TGT and a service ticket, gateway=true. Then:
The call to the responseBuilder looks like something like that:
`responseBuilder.build(CoreAuthenticationTestUtils.getWebApplicationService("jwtservice"),
                "ST-123456",
                CoreAuthenticationTestUtils.getAuthentication());` because there is a service ticket.
And as tokenAsResponse=true=> let's generate the JWT from the service ticket.

Scenario 4: The service is configured to send the  service ticket as a JWT and there is no SSO session(no TGT, no Service ticket), gateway=true. Then:
The call to the responseBuilder looks like something like that:
`responseBuilder.build(CoreAuthenticationTestUtils.getWebApplicationService("jwtservice"),
                "",
                CoreAuthenticationTestUtils.getAuthentication());` because there is no service ticket.
And as tokenAsResponse=true=> CAS tries to generate the JWT from the service ticket. But there is no service ticket so it ends with an exception.

That's why the parameter gateway is not there. And that's why I renamed to unit test. It's not really related to the parameter gateway. It's related to the fact that there might be no service ticket.